crates/oxc_sourcemap/src/encode.rs

```diff
@@ -45,43 +49,77 @@ pub fn encode_to_string(sourcemap: &SourceMap) -> Result<String> {
         contents.push("\",".into());
     }
     contents.push("\"names\":[".into());
-    for n in &sourcemap.names {
-        contents.push(serde_json::to_string(n.as_ref())?.into());
-        contents.push(",".into());
-    }
     if !sourcemap.names.is_empty() {
+        for n in &sourcemap.names {
+            if needs_escaping(n) {
+                contents.push(serde_json::to_string(n.as_ref())?.into());
+            } else {
+                contents.push("\"".into());
+                contents.push((&**n).into());
+                contents.push("\"".into());
+            }
+            contents.push(",".into());
+        }
         // Remove the last `,`.
         contents.pop();
     }
     contents.push(Cow::Borrowed("],\"sources\":["));
-    for s in &sourcemap.sources {
-        contents.push(serde_json::to_string(s.as_ref())?.into());
-        contents.push(",".into());
-    }
     if !sourcemap.sources.is_empty() {
+        for s in &sourcemap.sources {
+            if needs_escaping(s) {
+                contents.push(serde_escape_content(s)?.into());
+            } else {
+                contents.push("\"".into());
+                contents.push((&**s).into());
+                contents.push("\"".into());
+            }
+            contents.push(",".into());
+        }
         // Remove the last `,`.
         contents.pop();
     }
     // Quote `source_content` at parallel.
     if let Some(source_contents) = &sourcemap.source_contents {
         contents.push("],\"sourcesContent\":[".into());
-        cfg_if::cfg_if! {
-            if #[cfg(feature = "concurrent")] {
-                let quote_source_contents = source_contents
+        if !source_contents.is_empty() {
+            #[cfg(feature = "concurrent")]
+            {
+                let content_len = AtomicUsize::new(0);
+                let mut quote_source_contents = source_contents
                     .par_iter()
-                    .map(|x| serde_json::to_string(x.as_ref()))
-                    .collect::<std::result::Result<Vec<_>, serde_json::Error>>()
-                    .map_err(Error::from)?;
-            } else {
-                let quote_source_contents = source_contents
-                    .iter()
-                    .map(|x| serde_json::to_string(x.as_ref()))
-                    .collect::<std::result::Result<Vec<_>, serde_json::Error>>()
-                    .map_err(Error::from)?;
+                    .filter_map(|s| {
+                        if needs_escaping(s) {
```

Comment:
Can probably omit `needs_escaping(s)` call here. It'd be extremely uncommon for source content not to include at least 1 line break (which needs escaping), so it will basically always return true.